### ISSUE TITLE
Delay in transforming RealmResults to local instances (replica of RealmObjects) using Kotlin's map() function

### ISSUE THREAD

**Issue by PriyaSindkar** [realm/realm-kotlin#1331]
We am currently working on an android app which uses realm database. We are using architecture which abstracts RealmObjects obtained from Realm queries and converting them to local instances (standalone model replicating RealmObjects data).

The problem is- When we have just 550 RealmResults (with several nested objects and nested RealmResults inside each item) and we try to use `map()` function to transform the RealmObjects to the local models there is ~3s delay in transformation done via `map()`. This is a serious performance issue we are facing with only 550 rows of the parent RealmResults list.
Using `copyFromRealm()` and directly using this `List<RealmObject>` reduces the delay to ~1s but as the rows in db will increase, so will this delay.

Can you please suggest a better way to reduce this delay and keep using the local model instances?

Example of the RealmObject s we have in our database-

```
class Animals (val dogs : RealmList<Dog>, ...) : RealmObject

class Dog (val retrivers: RealmList<Retriever>, val features: Features>, ...) : RealmObject

class Retriever (val breed: RealmList<Breed>, val featuresMetaData: FeaturesMetaData>, ...): RealmObject

class Breed (val name: String, val featuresMetaData1: FeaturesMetaData1, ...): RealmObject

class Features (val name: String, val someRealmObject1: SomeRealmObject1, val someRealmObject2: SomeRealmObject2, val someRealmObject3: SomeRealmObject3, ...>: RealmObject

class FeaturesMetaData (val name: String, ...>: RealmObject

class FeaturesMetaData1 (val name: String, ...>: RealmObject
```
Basically, we are having a lot of internal dependencies. Our use case is that we want to retrive all the rows from Animals tables and then derive some calculations from this class and the inner classes it has. And since we need to do some calculations on all the data fetched, pagination in Realm itself is not an option.

**Comment by cmelchior** [comment]
Hi @PriyaSindkar 

Even though it is quite a deep schema hierarchy you have there, 3 seconds sounds like quite a lot. Can you post the code on how exactly you copy the data out and how you measure it?

Also, you can try to play around with the `depth` parameter to `copyFromRealm`. I know it will not fix your problem, but it could point to where the problem is, i.e what happens if you set the depth to `0` ?

**Comment by PriyaSindkar** [comment]
Hi @cmelchior,
tried your solution but no luck in performance. Tried with 2 ways- 
1. Defining the depth to 0 and then gradually fetching data when required (again with depth = 0) and
2. Without setting depth to 0 and fetching all records from the entire heirarchy.

My observation is this- When there is ~12500 rows of data in the child node table, `copyFromRealm()` is taking time. To use `copyFromRealm()` on this data took almost 900ms. With increase in data, this delay will also increase.

**Comment by cmelchior** [comment]
I wasn't expected using the `depth` to fix it, but it was more to get a sense of where the problem is. In the first post, you mentioned that it took 3 seconds for 550 objects. Are you saying that with 12500 objects it now takes ~1 second? Or was that with changing the depth?

No matter what, copying out a large number of objects will take time, I.e. I wouldn't consider copying out ~13k objects in 1 second as unreasonable slow, whereas 500 objects in 3 seconds would be very slow.

Can you post how you are measuring the performance of this?

**Comment by PriyaSindkar** [comment]
Using the example above, I have 550 animals and ~12500 breeds. For measuring performance I am simply using `measureTimeInMillis {}` .. wrapping it around `copyToRealm()`. All this is without using depth

**Comment by rorbech** [comment]
As mentioned, if you are in fact copy the whole database into memory in new objects it will take time. If you are copying the whole datamodel you are not only copying 550 animals and 12500 breeds, but also all the `Features`, `FeatureMetadata`, etc. Do you have an indicator of the total number of objects copied out to get an indicator of whether 3 seconds sound unreasonable?

Further, how are you copying the data? There are various `copyFromRealm`-method in our API, but if you call it on individual top level objects then the child objects will be replicated multiple times if the same instance is referenced in multiple top level hierarchies. To ensure that you only get one copy of each object you should be certain to call `copyFromRealm` once on some top level graph entry point. That can be either by modelling you data with a single parent or by calling `copyFromRealm` on one a collection type, i.e. `RealmList`, `RealmSet` or maybe even more interesting directly on `RealmResult` returned from a query. This allows us to maintain a cache while copying objects and avoid duplicates and use Kotlin references between the objects. 

Ccould you explain what kind of calculations that you are doing that require to have all objects in memory at a time? We have an issue around the concept of [projections ](https://github.com/realm/realm-kotlin/issues/947) and it would be nice to have insight in your use case, if it was something that we should have in mind for that.

**Comment by PriyaSindkar** [comment]
@rorbech - I am calling `copyFromRealm()` on top level graph entry point, i.e., taking the above example - Calling `copyFromRealm()` on  `RealmResults<Animals>`. 

And then approaching calculations in either of the 2 ways-
1. Doing calculations on these copies (without mapping to local instances)
2. Doing calculations on local instances (with mapping these copies to local instances)

Both of the above are taking almost same time overall. The 1st approach takes time with the calculations and the 2nd approach takes time in mapping.

For the 1st approach, I am calculating something like this-

```
val animals = fetchedAnimalsFromDB().copyFromRealm()
animals.performCalculationsOnRealmInstances()
```

For the 2nd approach, I am mapping all the classes in the heirarchy individually like so-

```
val animals = fetchedAnimalsFromDB().copyFromRealm().map { LocalAnimals.init(it) }
animals.performCalculations()

... 

// Sample model class
class LocalAnimals() {
...
  companion object {
      fun init(animal: Animal) {
            variable1 = animal.variable1,
            variable2 = animal.variable2,
            dogs = animal.dogs.map { Dog.init(it) }
      }
  }  
}

```

The calculations are probablities, averages, and percentage calculations that are done on various Integers gotten from the child classes like `breeds` and `featuresMetaData1` . for this I am using various kotlin functions like- `map()`, `flatMap()`, `maxOf()`, `sumOf()`, etc. This is not much data (sorry cannot share the actual code which would help you certainly) but I hope it helps you a bit with the projections.
